ec/internal/planmodifiers/use_state_unless_template_changed.go

```diff
@@ -20,42 +20,72 @@ package planmodifiers
 import (
 	"context"
 
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
 )
 
 // Use current state for a topology's attribute if the topology's state is not nil and the template attribute has not changed
-func UseStateForUnknownUnlessTemplateChanged() planmodifier.String {
-	return useStateForUnknownUnlessTemplateChanged{}
+func UseStateForUnknownUnlessMigrationisRequired(resourceKind string) useStateForUnknownUnlessTemplateChanged {
+	return useStateForUnknownUnlessTemplateChanged{resourceKind: resourceKind}
 }
 
-type useStateForUnknownUnlessTemplateChanged struct{}
+type useStateForUnknownUnlessTemplateChanged struct {
+	resourceKind string
+}
+
+type PlanModifierResponse interface {
+	planmodifier.StringResponse | planmodifier.Int64Response
+}
 
 func (m useStateForUnknownUnlessTemplateChanged) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
-	if !resp.PlanValue.IsUnknown() {
-		return
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
 
-	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
-	if req.ConfigValue.IsUnknown() {
-		return
+func (m useStateForUnknownUnlessTemplateChanged) PlanModifyInt64(ctx context.Context, req planmodifier.Int64Request, resp *planmodifier.Int64Response) {
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
+
+func (m useStateForUnknownUnlessTemplateChanged) UseState(ctx context.Context, configValue attr.Value, plan tfsdk.Plan, state tfsdk.State, planValue attr.Value) (bool, diag.Diagnostics) {
+	var diags diag.Diagnostics
 
-	if req.StateValue.IsNull() {
-		return
+	if !planValue.IsUnknown() {
+		return false, nil
 	}
 
-	templateChanged, diags := AttributeChanged(ctx, path.Root("deployment_template_id"), req.Plan, req.State)
-	resp.Diagnostics.Append(diags...)
+	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
+	if configValue.IsUnknown() {
+		return false, nil
+	}
+
+	templateChanged, d := AttributeChanged(ctx, path.Root("deployment_template_id"), plan, state)
+	diags.Append(d...)
+
+	var migrateToLatestHw bool
```

Comment:
It should be a TF type I think - the attribute's value can be `Unknown`.

Reply:
I was assuming that `unknown` would read as `false`, which is the behaviour we want in case the field is not specified in the TF config 🤔 Does this make sense?

Reply:
Yes, it makes sense. But while it works here, we still rely on the TF framework implementation to read "unknown" state into a type that doesn't support this.

Reply:
We can return from func here, if template changed.